pkg/metrics/osbclientproxy/osbproxy.go

```diff
@@ -0,0 +1,189 @@
+func (pc proxyclient) DeprovisionInstance(r *osb.DeprovisionRequest) (*osb.DeprovisionResponse, error) {
+	glog.V(9).Info("OSBClientProxy DeprovisionInstance()")
+	response, err := pc.realOSBClient.DeprovisionInstance(r)
+	pc.updateMetrics(deprovisionInstance, err)
+	return response, err
+}
+
+// PollLastOperation implements
+// go-open-service-broker-client/v2/Client.PollLastOperation by proxying the
+// method to the underlying implementation and capturing request metrics.
+func (pc proxyclient) PollLastOperation(r *osb.LastOperationRequest) (*osb.LastOperationResponse, error) {
+	glog.V(9).Info("OSBClientProxy PollLastOperation()")
+	response, err := pc.realOSBClient.PollLastOperation(r)
+	pc.updateMetrics(pollLastOperation, err)
+	return response, err
+}
+
+// PollBindingLastOperation implements
+// go-open-service-broker-client/v2/Client.PollBindingLastOperation by proxying
+// the method to the underlying implementation and capturing request metrics.
+func (pc proxyclient) PollBindingLastOperation(r *osb.BindingLastOperationRequest) (*osb.LastOperationResponse, error) {
+	glog.V(9).Info("OSBClientProxy PollBindingLastOperation()")
+	response, err := pc.realOSBClient.PollBindingLastOperation(r)
+	pc.updateMetrics(pollBindingLastOperation, err)
+	return response, err
+}
+
+// Bind implements go-open-service-broker-client/v2/Client.Bind by proxying the
+// method to the underlying implementation and capturing request metrics.
+func (pc proxyclient) Bind(r *osb.BindRequest) (*osb.BindResponse, error) {
+	glog.V(9).Info("OSBClientProxy Bind().")
+	response, err := pc.realOSBClient.Bind(r)
+	pc.updateMetrics(bind, err)
+	return response, err
+}
+
+// Unbind implements go-open-service-broker-client/v2/Client.Unbind by proxying
+// the method to the underlying implementation and capturing request metrics.
+func (pc proxyclient) Unbind(r *osb.UnbindRequest) (*osb.UnbindResponse, error) {
+	glog.V(9).Info("OSBClientProxy Unbind()")
+	response, err := pc.realOSBClient.Unbind(r)
+	pc.updateMetrics(unbind, err)
+	return response, err
+}
+
+// GetBinding implements go-open-service-broker-client/v2/Client.GetBinding by
+// proxying the method to the underlying implementation and capturing request
+// metrics.
+func (pc proxyclient) GetBinding(r *osb.GetBindingRequest) (*osb.GetBindingResponse, error) {
+	glog.V(9).Info("OSBClientProxy GetBinding()")
+	response, err := pc.realOSBClient.GetBinding(r)
+	pc.updateMetrics(getBinding, err)
+	return response, err
+}
+
+const (
+	oneXX     = "1xx"
+	twoXX     = "2xx"
+	threeXX   = "3xx"
+	fourXX    = "4xx"
+	fiveXX    = "5xx"
+	clientErr = "client-error"
+)
+
+// updateMetrics bumps the request count metric for the specific broker, method
+// and status
+func (pc proxyclient) updateMetrics(method string, err error) {
+	var statusGroup string
+
+	if err == nil {
+		metrics.OSBRequestCount.WithLabelValues(pc.brokerName, method, twoXX).Inc()
+		return
+	}
+
+	status, ok := osb.IsHTTPError(err)
```

Comment:
Something seems off to me. If this only takes the `err` how are getting the response code? Does the osbclient always send a non-nil err even on things like 200?

Reply:
The response code is in the error when the error is an HTTPStatusCodeError. When `err` is nil, the status code is assumed to be something in the 2xx family.

Can you be some specific about what you think is wrong?

Reply:
Is there confusion about what the response is? The response from the osb client is not an HTTP response. It is the body from the HTTP response. For example, when fetching the catalog, the response would be a `CatalogResponse`.